File: py/indicators/marc_chaikin/advance_decline_oscillator/advance_decline_oscillator.py
"""Marc Chaikin's Advance-Decline Oscillator (ADOSC)."""

import math

from .params import AdvanceDeclineOscillatorParams, MovingAverageType
from ...common.simple_moving_average.simple_moving_average import SimpleMovingAverage
from ...common.simple_moving_average.params import SimpleMovingAverageParams
from ...common.exponential_moving_average.exponential_moving_average import ExponentialMovingAverage
from ...common.exponential_moving_average.params import ExponentialMovingAverageLengthParams
from ...core.line_indicator import LineIndicator
from ...core.indicator import Indicator
from ...core.identifier import Identifier
from ...core.metadata import Metadata
from ...core.build_metadata import build_metadata, OutputText
from ...core.output import Output
from ....entities.bar import Bar
from ....entities.quote import Quote
from ....entities.trade import Trade
from ....entities.scalar import Scalar
from ....entities.bar_component import DEFAULT_BAR_COMPONENT, bar_component_value
from ....entities.quote_component import DEFAULT_QUOTE_COMPONENT, quote_component_value
from ....entities.trade_component import DEFAULT_TRADE_COMPONENT, trade_component_value


class AdvanceDeclineOscillator(Indicator):
    """Marc Chaikin's Advance-Decline Oscillator (ADOSC).

    ADOSC = FastMA(AD) - SlowMA(AD)

    where AD is the cumulative Advance-Decline line:
    CLV = ((Close - Low) - (High - Close)) / (High - Low)
    AD = AD_prev + CLV * Volume
    """

    def __init__(self, params: AdvanceDeclineOscillatorParams) -> None:
        if params.fast_length < 2:
            raise ValueError("invalid advance-decline oscillator parameters: "
                             "fast length should be greater than 1")
        if params.slow_length < 2:
            raise ValueError("invalid advance-decline oscillator parameters: "
                             "slow length should be greater than 1")

        if params.moving_average_type == MovingAverageType.SMA:
            ma_label = "SMA"
            self._fast_ma = SimpleMovingAverage(
                SimpleMovingAverageParams(length=params.fast_length))
            self._slow_ma = SimpleMovingAverage(
                SimpleMovingAverageParams(length=params.slow_length))
        else:
            ma_label = "EMA"
            self._fast_ma = ExponentialMovingAverage.from_length(
                ExponentialMovingAverageLengthParams(
                    length=params.fast_length,
                    first_is_average=params.first_is_average))
            self._slow_ma = ExponentialMovingAverage.from_length(
                ExponentialMovingAverageLengthParams(
                    length=params.slow_length,
                    first_is_average=params.first_is_average))

        mnemonic = f"adosc({ma_label}{params.fast_length}/{ma_label}{params.slow_length})"
        description = "Chaikin Advance-Decline Oscillator " + mnemonic

        bar_func = bar_component_value(DEFAULT_BAR_COMPONENT)
        quote_func = quote_component_value(DEFAULT_QUOTE_COMPONENT)
        trade_func = trade_component_value(DEFAULT_TRADE_COMPONENT)

        self._line = LineIndicator(mnemonic, description, bar_func, quote_func, trade_func, self.update)
        self._ad = 0.0
        self._value = math.nan
        self._primed = False

    def is_primed(self) -> bool:
        return self._primed

    def metadata(self) -> Metadata:
        return build_metadata(
            Identifier.ADVANCE_DECLINE_OSCILLATOR,
            self._line.mnemonic,
            self._line.description,
            [OutputText(self._line.mnemonic, self._line.description)],
        )

    def update(self, sample: float) -> float:
        """Update with scalar (H=L=C, volume=1, AD unchanged but fed to MAs)."""
        if math.isnan(sample):
            return math.nan
        return self.update_hlcv(sample, sample, sample, 1.0)

    def update_hlcv(self, high: float, low: float, close: float, volume: float) -> float:
        """Update with high, low, close, volume."""
        if math.isnan(high) or math.isnan(low) or math.isnan(close) or math.isnan(volume):
            return math.nan

        temp = high - low
        if temp > 0:
            self._ad += ((close - low) - (high - close)) / temp * volume

        fast = self._fast_ma.update(self._ad)
        slow = self._slow_ma.update(self._ad)
        self._primed = self._fast_ma.is_primed() and self._slow_ma.is_primed()

        if math.isnan(fast) or math.isnan(slow):
            self._value = math.nan
            return self._value

        self._value = fast - slow
        return self._value

    def update_scalar(self, sample: Scalar) -> Output:
        return self._line.update_scalar(sample)

    def update_bar(self, sample: Bar) -> Output:
        """Shadows LineIndicator.update_bar to extract HLCV."""
        value = self.update_hlcv(sample.high, sample.low, sample.close, sample.volume)
        return [Scalar(time=sample.time, value=value)]

    def update_quote(self, sample: Quote) -> Output:
        return self._line.update_quote(sample)

    def update_trade(self, sample: Trade) -> Output:
        return self._line.update_trade(sample)

File: rs/src/indicators/marc_chaikin/advance_decline_oscillator/advance_decline_oscillator.rs
use std::any::Any;

use crate::entities::bar::Bar;
use crate::entities::quote::Quote;
use crate::entities::scalar::Scalar;
use crate::entities::trade::Trade;
use crate::indicators::common::exponential_moving_average::{
    ExponentialMovingAverage, ExponentialMovingAverageLengthParams,
};
use crate::indicators::common::simple_moving_average::{
    SimpleMovingAverage, SimpleMovingAverageParams,
};
use crate::indicators::core::build_metadata::{build_metadata, OutputText};
use crate::indicators::core::identifier::Identifier;
use crate::indicators::core::indicator::Indicator;
use crate::indicators::core::metadata::Metadata;

/// Output index for AdvanceDeclineOscillator.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum AdvanceDeclineOscillatorOutput {
    Value = 1,
}

/// Moving average type selector.
#[repr(u8)]
#[derive(Debug, Clone, Copy, PartialEq)]
pub enum AdvanceDeclineOscillatorMaType {
    SMA = 0,
    EMA = 1,
}

/// Parameters for AdvanceDeclineOscillator.
#[derive(Debug, Clone)]
pub struct AdvanceDeclineOscillatorParams {
    pub fast_length: i64,
    pub slow_length: i64,
    // Note: SMA uses usize, EMA uses i64. Conversion handled in new().
    pub moving_average_type: AdvanceDeclineOscillatorMaType,
    pub first_is_average: bool,
}

impl Default for AdvanceDeclineOscillatorParams {
    fn default() -> Self {
        Self {
            fast_length: 3,
            slow_length: 10,
            moving_average_type: AdvanceDeclineOscillatorMaType::EMA,
            first_is_average: false,
        }
    }
}

trait LineUpdater {
    fn update(&mut self, sample: f64) -> f64;
    fn is_primed(&self) -> bool;
}

struct SmaUpdater {
    inner: SimpleMovingAverage,
}

impl LineUpdater for SmaUpdater {
    fn update(&mut self, sample: f64) -> f64 {
        self.inner.update(sample)
    }
    fn is_primed(&self) -> bool {
        self.inner.is_primed()
    }
}

struct EmaUpdater {
    inner: ExponentialMovingAverage,
}

impl LineUpdater for EmaUpdater {
    fn update(&mut self, sample: f64) -> f64 {
        self.inner.update(sample)
    }
    fn is_primed(&self) -> bool {
        self.inner.is_primed()
    }
}

/// Marc Chaikin's Advance-Decline Oscillator.
pub struct AdvanceDeclineOscillator {
    ad: f64,
    fast_ma: Box<dyn LineUpdater>,
    slow_ma: Box<dyn LineUpdater>,
    value: f64,
    primed: bool,
}

impl AdvanceDeclineOscillator {
    pub fn new(params: &AdvanceDeclineOscillatorParams) -> Result<Self, String> {
        if params.fast_length < 2 {
            return Err(
                "invalid advance-decline oscillator parameters: fast length should be greater than 1"
                    .to_string(),
            );
        }
        if params.slow_length < 2 {
            return Err(
                "invalid advance-decline oscillator parameters: slow length should be greater than 1"
                    .to_string(),
            );
        }

        let (fast_ma, slow_ma): (Box<dyn LineUpdater>, Box<dyn LineUpdater>) =
            match params.moving_average_type {
                AdvanceDeclineOscillatorMaType::SMA => {
                    let fast = SimpleMovingAverage::new(&SimpleMovingAverageParams {
                        length: params.fast_length as usize,
                        ..Default::default()
                    })?;
                    let slow = SimpleMovingAverage::new(&SimpleMovingAverageParams {
                        length: params.slow_length as usize,
                        ..Default::default()
                    })?;
                    (
                        Box::new(SmaUpdater { inner: fast }),
                        Box::new(SmaUpdater { inner: slow }),
                    )
                }
                AdvanceDeclineOscillatorMaType::EMA => {
                    let fast = ExponentialMovingAverage::new_from_length(
                        &ExponentialMovingAverageLengthParams {
                            length: params.fast_length,
                            first_is_average: params.first_is_average,
                            ..Default::default()
                        },
                    )?;
                    let slow = ExponentialMovingAverage::new_from_length(
                        &ExponentialMovingAverageLengthParams {
                            length: params.slow_length,
                            first_is_average: params.first_is_average,
                            ..Default::default()
                        },
                    )?;
                    (
                        Box::new(EmaUpdater { inner: fast }),
                        Box::new(EmaUpdater { inner: slow }),
                    )
                }
            };

        Ok(Self {
            ad: 0.0,
            fast_ma,
            slow_ma,
            value: f64::NAN,
            primed: false,
        })
    }

    /// Update with high, low, close, volume.
    pub fn update_hlcv(&mut self, high: f64, low: f64, close: f64, volume: f64) -> f64 {
        if high.is_nan() || low.is_nan() || close.is_nan() || volume.is_nan() {
            return f64::NAN;
        }

        let temp = high - low;
        if temp > 0.0 {
            self.ad += ((close - low) - (high - close)) / temp * volume;
        }

        let fast = self.fast_ma.update(self.ad);
        let slow = self.slow_ma.update(self.ad);
        self.primed = self.fast_ma.is_primed() && self.slow_ma.is_primed();

        if fast.is_nan() || slow.is_nan() {
            self.value = f64::NAN;
            return self.value;
        }

        self.value = fast - slow;
        self.value
    }
}

impl Indicator for AdvanceDeclineOscillator {
    fn is_primed(&self) -> bool {
        self.primed
    }

    fn metadata(&self) -> Metadata {
        build_metadata(
            Identifier::AdvanceDeclineOscillator,
            "adosc",
            "Chaikin Advance-Decline Oscillator",
            &[OutputText {
                mnemonic: "adosc".to_string(),
                description: "Chaikin Advance-Decline Oscillator".to_string(),
            }],
        )
    }

    fn update_scalar(&mut self, scalar: &Scalar) -> Vec<Box<dyn Any>> {
        let v = self.update_hlcv(scalar.value, scalar.value, scalar.value, 1.0);
        vec![Box::new(Scalar::new(scalar.time, v))]
    }

    fn update_bar(&mut self, bar: &Bar) -> Vec<Box<dyn Any>> {
        let v = self.update_hlcv(bar.high, bar.low, bar.close, bar.volume);
        vec![Box::new(Scalar::new(bar.time, v))]
    }

    fn update_quote(&mut self, quote: &Quote) -> Vec<Box<dyn Any>> {
        let mid = quote.mid();
        let v = self.update_hlcv(mid, mid, mid, 1.0);
        vec![Box::new(Scalar::new(quote.time, v))]
    }

    fn update_trade(&mut self, trade: &Trade) -> Vec<Box<dyn Any>> {
        let v = self.update_hlcv(trade.price, trade.price, trade.price, 1.0);
        vec![Box::new(Scalar::new(trade.time, v))]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use super::super::testdata::testdata;
    fn round_to(v: f64, digits: i32) -> f64 {
        let p = 10f64.powi(digits);
        (v * p).round() / p
    }

    #[test]
    fn test_advance_decline_oscillator_ema() {
        let highs = testdata::test_highs();
        let lows = testdata::test_lows();
        let closes = testdata::test_closes();
        let volumes = testdata::test_volumes();
        let expected = testdata::test_expected_ema();

        let mut adosc = AdvanceDeclineOscillator::new(&AdvanceDeclineOscillatorParams {
            fast_length: 3,
            slow_length: 10,
            moving_average_type: AdvanceDeclineOscillatorMaType::EMA,
            first_is_average: false,
        })
        .unwrap();

        for i in 0..highs.len() {
            let v = adosc.update_hlcv(highs[i], lows[i], closes[i], volumes[i]);

            if i < 9 {
                assert!(v.is_nan(), "[{}] expected NaN, got {}", i, v);
                assert!(!adosc.is_primed(), "[{}] expected not primed", i);
                continue;
            }

            assert!(!v.is_nan(), "[{}] expected non-NaN", i);
            assert!(adosc.is_primed(), "[{}] expected primed", i);

            let got = round_to(v, 2);
            let exp = round_to(expected[i], 2);
            assert_eq!(got, exp, "[{}] expected {}, got {}", i, exp, got);
        }
    }

    #[test]
    fn test_advance_decline_oscillator_sma() {
        let highs = testdata::test_highs();
        let lows = testdata::test_lows();
        let closes = testdata::test_closes();
        let volumes = testdata::test_volumes();
        let expected = testdata::test_expected_sma();

        let mut adosc = AdvanceDeclineOscillator::new(&AdvanceDeclineOscillatorParams {
            fast_length: 3,
            slow_length: 10,
            moving_average_type: AdvanceDeclineOscillatorMaType::SMA,
            first_is_average: false,
        })
        .unwrap();

        for i in 0..highs.len() {
            let v = adosc.update_hlcv(highs[i], lows[i], closes[i], volumes[i]);

            if i < 9 {
                assert!(v.is_nan(), "[{}] expected NaN, got {}", i, v);
                assert!(!adosc.is_primed(), "[{}] expected not primed", i);
                continue;
            }

            assert!(!v.is_nan(), "[{}] expected non-NaN", i);
            assert!(adosc.is_primed(), "[{}] expected primed", i);

            let got = round_to(v, 2);
            let exp = round_to(expected[i], 2);
            assert_eq!(got, exp, "[{}] expected {}, got {}", i, exp, got);
        }
    }

    #[test]
    fn test_advance_decline_oscillator_invalid_params() {
        let r = AdvanceDeclineOscillator::new(&AdvanceDeclineOscillatorParams {
            fast_length: 1,
            slow_length: 10,
            ..Default::default()
        });
        assert!(r.is_err());

        let r = AdvanceDeclineOscillator::new(&AdvanceDeclineOscillatorParams {
            fast_length: 3,
            slow_length: 1,
            ..Default::default()
        });
        assert!(r.is_err());
    }
}
